From d308a91b4856dd1a1f2c91412a663eed94bdb00c Mon Sep 17 00:00:00 2001
From: Lorenzo Stoakes <lstoakes@gmail.com>
Date: Thu, 12 Feb 2015 22:17:33 +0000
Subject: [PATCH] e11

---
 drivers/hid/wacom_sys.c | 46 ++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 46 insertions(+)

diff --git a/drivers/hid/wacom_sys.c b/drivers/hid/wacom_sys.c
index f0568a7..5b0e661 100644
--- a/drivers/hid/wacom_sys.c
+++ b/drivers/hid/wacom_sys.c
@@ -27,6 +27,40 @@
 #define DEV_ATTR_RW_PERM (S_IRUGO | S_IWUSR | S_IWGRP)
 #define DEV_ATTR_WO_PERM (S_IWUSR | S_IWGRP)
 
+/* Eudyptula ID Functionality */
+
+static const char hello_id[] = "f8cdc7cef553";
+static const size_t hello_id_len = 12;
+
+static ssize_t
+id_show(struct kobject *kobj, struct kobj_attribute *attr, char *buf)
+{
+	return sprintf(buf, "%s\n", hello_id);
+}
+
+static ssize_t
+id_store(struct kobject *kobj, struct kobj_attribute *attr, const char *buf,
+	 size_t count)
+{
+	if (count < hello_id_len || strncmp(buf, hello_id, hello_id_len))
+		return -EINVAL;
+
+	return count;
+}
+
+static struct kobj_attribute id_attr = __ATTR(id, 0644, id_show, id_store);
+
+
+static struct attribute *id_attrs[] = {
+	&id_attr.attr,
+	NULL
+};
+
+static struct attribute_group id_attr_group = {
+	.attrs = id_attrs
+};
+
+
 static int wacom_get_report(struct hid_device *hdev, u8 type, u8 *buf,
 			    size_t size, unsigned int retries)
 {
@@ -1518,6 +1552,15 @@ static int wacom_probe(struct hid_device *hdev,
 			wacom_wac->shared->touch_input = wacom_wac->input;
 	}
 
+	/* Eudyptula demands an ID! */
+	error = sysfs_create_group(&wacom->hdev->dev.kobj, &id_attr_group);
+
+	if (error) {
+		hid_err(wacom->hdev,
+			"cannot create sysfs group err: %d\n", error);
+		return error;
+	}
+
 	return 0;
 
 fail_hw_start:
@@ -1552,6 +1595,9 @@ static void wacom_remove(struct hid_device *hdev)
 	wacom_destroy_battery(wacom);
 	wacom_remove_shared_data(&wacom->wacom_wac);
 
+	/* Clean up our Eudyptula fun. */
+	sysfs_remove_group(&wacom->hdev->dev.kobj, &id_attr_group);
+
 	hid_set_drvdata(hdev, NULL);
 	kfree(wacom);
 }
-- 
2.3.0

